<a href="https://colab.research.google.com/github/LiTaokui/TensorFlow_Toturial/blob/main/Dropout_EX_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [24]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [25]:
import warnings
warnings.filterwarnings('ignore')

In [26]:
df = pd.read_csv("/content/sample_data/sonar_dataset.csv", header=None)
df.sample(5)

0       1       2       3       4       5       6       7       8   \
119  0.0261  0.0266  0.0223  0.0749  0.1364  0.1513  0.1316  0.1654  0.1864   
151  0.0231  0.0315  0.0170  0.0226  0.0410  0.0116  0.0223  0.0805  0.2365   
142  0.0526  0.0563  0.1219  0.1206  0.0246  0.1022  0.0539  0.0439  0.2291   
68   0.0195  0.0142  0.0181  0.0406  0.0391  0.0249  0.0892  0.0973  0.0840   
196  0.0050  0.0017  0.0270  0.0450  0.0958  0.0830  0.0879  0.1220  0.1977   

         9   ...      51      52      53      54      55      56      57  \
119  0.2013  ...  0.0222  0.0175  0.0127  0.0022  0.0124  0.0054  0.0021   
151  0.2461  ...  0.0125  0.0036  0.0123  0.0043  0.0114  0.0052  0.0091   
142  0.1632  ...  0.0339  0.0149  0.0335  0.0376  0.0174  0.0132  0.0103   
68   0.1191  ...  0.0028  0.0067  0.0120  0.0012  0.0022  0.0058  0.0042   
196  0.2282  ...  0.0165  0.0056  0.0010  0.0027  0.0062  0.0024  0.0063   

         58      59  60  
119  0.0028  0.0023   M  
151  0.0008  0.0092   M  
142  0.0364  0.0208   M  
68   0.0067  0.0012   R  
196  0.0017  0.0028   M  

[5 rows x 61 columns]

In [27]:
df.shape

(208, 61)

In [28]:
# check for nan values
df.isna().sum()

0     0
1     0
2     0
3     0
4     0
     ..
56    0
57    0
58    0
59    0
60    0
Length: 61, dtype: int64

In [29]:
df.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57, 58, 59, 60],
           dtype='int64')

In [30]:
df[60].value_counts() # label is not skewed

M    111
R     97
Name: 60, dtype: int64

In [31]:
X = df.drop(60, axis=1)
y = df[60]
y.head()

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object

In [32]:
y = pd.get_dummies(y, drop_first=True)
y.sample(5) # R --> 1 and M --> 0

R
171  0
95   1
68   1
74   1
153  0

In [33]:
y.value_counts()

R
0    111
1     97
dtype: int64

In [34]:
X.head()

0       1       2       3       4       5       6       7       8   \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

       9   ...      50      51      52      53      54      55      56  \
0  0.2111  ...  0.0232  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180   
1  0.2872  ...  0.0125  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140   
2  0.6194  ...  0.0033  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316   
3  0.1264  ...  0.0241  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050   
4  0.4459  ...  0.0156  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072   

       57      58      59  
0  0.0084  0.0090  0.0032  
1  0.0049  0.0052  0.0044  
2  0.0164  0.0095  0.0078  
3  0.0044  0.0040  0.0117  
4  0.0048  0.0107  0.0094  

[5 rows x 60 columns]

In [35]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [36]:
X_train.head()

0       1       2       3       4       5       6       7       8   \
67   0.0368  0.0403  0.0317  0.0293  0.0820  0.1342  0.1161  0.0663  0.0155   
14   0.0124  0.0433  0.0604  0.0449  0.0597  0.0355  0.0531  0.0343  0.1052   
164  0.0163  0.0198  0.0202  0.0386  0.0752  0.1444  0.1487  0.1484  0.2442   
179  0.0394  0.0420  0.0446  0.0551  0.0597  0.1416  0.0956  0.0802  0.1618   
19   0.0126  0.0149  0.0641  0.1732  0.2565  0.2559  0.2947  0.4110  0.4983   

         9   ...      50      51      52      53      54      55      56  \
67   0.0506  ...  0.0058  0.0091  0.0160  0.0160  0.0081  0.0070  0.0135   
14   0.2120  ...  0.0078  0.0083  0.0057  0.0174  0.0188  0.0054  0.0114   
164  0.2822  ...  0.0027  0.0077  0.0026  0.0031  0.0083  0.0020  0.0084   
179  0.2558  ...  0.0118  0.0146  0.0040  0.0114  0.0032  0.0062  0.0101   
19   0.5920  ...  0.0153  0.0092  0.0035  0.0098  0.0121  0.0006  0.0181   

         57      58      59  
67   0.0067  0.0078  0.0068  
14   0.0196  0.0147  0.0062  
164  0.0108  0.0083  0.0033  
179  0.0068  0.0053  0.0087  
19   0.0094  0.0116  0.0063  

[5 rows x 60 columns]

## Wihout Dropout ###

In [37]:
import tensorflow as tf
from tensorflow import keras

In [38]:
model = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dense(1, activation='sigmoid')
])

model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.6994 - accuracy: 0.4744
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6864 - accuracy: 0.5513
Epoch 3/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6758 - accuracy: 0.5449
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6619 - accuracy: 0.5833
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6420 - accuracy: 0.6859
Epoch 6/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6152 - accuracy: 0.6987
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5877 - accuracy: 0.7244
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5516 - accuracy: 0.7436
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.5164 - accuracy: 0.8077
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.4839 - accuracy: 0.8269
Epoch 11/

In [39]:
model.evaluate(X_test, y_test)

2/2 [==============================] - 0s 5ms/step - loss: 1.0192 - accuracy: 0.7692


[1.0192102193832397, 0.7692307829856873]

Training Accuracy >>> Test Accuracy

In [40]:
y_pred = model.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[2.9312508e-07 9.9490070e-01 9.7718859e-01 9.6019859e-05 9.9999762e-01
 9.9990261e-01 1.2114036e-01 9.9999946e-01 5.1198673e-05 9.9999905e-01]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [41]:
y_test[:10]

R
186  0
155  0
165  0
200  0
58   1
34   1
151  0
18   1
202  0
62   1

In [42]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.73      0.89      0.80        27
           1       0.84      0.64      0.73        25

    accuracy                           0.77        52
   macro avg       0.78      0.76      0.76        52
weighted avg       0.78      0.77      0.77        52



### With Dropout ###

In [43]:
modeld = keras.Sequential([
    keras.layers.Dense(60, input_dim=60, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(30, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(15, activation='relu'),
    keras.layers.Dropout(0.5),
    keras.layers.Dense(1, activation='sigmoid')
])

modeld.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

modeld.fit(X_train, y_train, epochs=100, batch_size=8)

Epoch 1/100
20/20 [==============================] - 1s 2ms/step - loss: 0.7453 - accuracy: 0.5449
Epoch 2/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7279 - accuracy: 0.4615
Epoch 3/100
20/20 [==============================] - 0s 3ms/step - loss: 0.6757 - accuracy: 0.5962
Epoch 4/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7099 - accuracy: 0.5128
Epoch 5/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6953 - accuracy: 0.5256
Epoch 6/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6985 - accuracy: 0.4487
Epoch 7/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6937 - accuracy: 0.5385
Epoch 8/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6991 - accuracy: 0.5385
Epoch 9/100
20/20 [==============================] - 0s 2ms/step - loss: 0.7072 - accuracy: 0.4808
Epoch 10/100
20/20 [==============================] - 0s 2ms/step - loss: 0.6688 - accuracy: 0.5833
Epoch 11/

In [44]:
modeld.evaluate(X_test, y_test)

2/2 [==============================] - 0s 5ms/step - loss: 0.4525 - accuracy: 0.7885


[0.4524887800216675, 0.7884615659713745]

In [45]:
y_pred = modeld.predict(X_test).reshape(-1)
print(y_pred[:10])

# round the values to nearest integer ie 0 or 1
y_pred = np.round(y_pred)
print(y_pred[:10])

[0.0015482  0.6138848  0.6894566  0.03307855 0.9998954  0.8498106
 0.2174339  0.9999058  0.05581111 0.99994373]
[0. 1. 1. 0. 1. 1. 0. 1. 0. 1.]


In [46]:
from sklearn.metrics import confusion_matrix , classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.75      0.89      0.81        27
           1       0.85      0.68      0.76        25

    accuracy                           0.79        52
   macro avg       0.80      0.78      0.78        52
weighted avg       0.80      0.79      0.79        52



You can see that by using dropout layer test accuracy increased from 0.77 to 0.79
 